design/one-pager-easy-stacks.md

```diff
@@ -0,0 +1,294 @@
+that we can improve during this process:
+* The preparation phase of these YAMLs requires to be very careful and changing
+  something usually means touching a few different files. While this may not seem
+  like a big hurdle, changing some high level parameters of that environment is
+  cumbersome _after_ it's deployed.
+* A user who is familiar with AWS may not be familiar with Azure or GCP; so,
+  ready-made configuration stacks that promise the exact same setup would be
+  useful for them to see how they can achieve similar setups and run various
+  benchmarks according to their business needs.
+* After you run `kubectl apply -f <directory of YAMLs>`, a dependency graph of
+  resources starts to resolve. However, it's not always clear for user to know _when_
+  everything is actually ready to use unless they know readiness of which resources
+  signal the readiness of the whole environment.
+
+## Goals
+
+The main goal of this design is to make it easier for people to write sets of
+configurations and be able to manage those configuration sets on a higher level.
+
+It is important that the design puts forward:
+
+* A base boilerplate package where creating a new configuration set that has a
+  controller is as easy as changing the YAML files.
+* A way to declare pre-requisites that needs to be satisfied by the cluster
+  for the configuration set to work.
+* A way to change a high level configuration by changing 1 value that is then
+  translated to the YAMLs that are already deployed.
+
+## Proposal
+
+Crossplane has the notion of stacks that can be used as extension to Crossplane's
+abilites, to deploy an application and all other things that you can do with
+Kubernetes controllers since main components of stacks are CRDs and controllers
+watching them.
+
+To achieve the mentioned goals, we can publish the YAML files in a stack with a
+controller that has its own CRD with high level parameters. Controller deploys
+and updates the YAMLs that are packaged in the stack image according to user's
+input on the CR instance of that stack that user creates.
+
+As sky is the limit when it comes to stacks, we need to put down some general
+rules to be adhered by initial configuration stacks.
+
+* CRDs of the resources that will be deployed by the stack are declared as
+  dependency so that user knows what CR instances will be deployed by the given
+  configuration stack.
+* It should be possible to have different sets of configurations in one stack
+  and user should be able to choose from them via the CR that is an instance
+  of stack's CRD.
+  * Note that it's highly preferred to have 1 custom resource definition that
+    the stack uses and expose the high level variations of the parameters through
+    `spec` instead of having different custom resource definitions for each class
+    of configuration.
+* A minimal high level set of configurations are exposed to the user through
```

Comment:
> if this will include passing the provider secret content via environment config CR

Not sure about passing down the secret _content_ via CR.

I think taking a `Secret` name and namespace in a field like `spec.credentialsSecretRef` is the easiest option. This way the client that will put the credentials doesn't have to know how to construct a specific `Provider` CR. Client will;
1. Create a secret.
2. Create a `EasyAWS` CR with `spec.credentialsSecretRef` that points to the created secret.

> requiredResources:
> resource:

I couldn't find a reference to these properties of `app.yaml`. Could you point me to a resource that explains how they work?

> Note that maxAllowed would limit the number of environment config CRs created, at least for now as a simplifying assumption, unless it would make sense to allow multiple env. config CRs for a given config stack in the first release.

Limiting the number of CRs to 1 would indeed simplify the initial implementation. But I'm not aware of a way to tell api-server not to accept creation call of specific CR after some point.

Reply:
> But I'm not aware of a way to tell api-server not to accept creation call of specific CR after some point.

As of Kubernetes 1.15 you can use a `ResourceQuota` object to [restrict the count](https://kubernetes.io/docs/concepts/policy/resource-quotas/#object-count-quota) of custom resource instances in a namespace. Maybe we could create `ResourceQuota` objects on stack installation based on the `maxAllowed` fields!

Reply:
> I think taking a Secret name and namespace in a field like spec.credentialsSecretRef is the easiest option.

I agree with this. Creating a secret separately and then putting a reference in the spec field of the single CR seems like it would be more manageable than requiring multiple resources (and coordinating that).

Reply:
> I think taking a Secret name and namespace in a field like spec.credentialsSecretRef is the easiest option.

Throwing my support behind this too - Secrets are the right place for credentials. Let's not include sensitive credentials inline in our custom resources; it shouldn't be too onerous to create a (provider credentials) `Secret` and an `EasyAWS` that references it in one fell swoop.

> Limiting the number of CRs to 1 would indeed simplify the initial implementation.

I can't recall ever coming across an example of a singleton resource in Kubernetes (e.g. only one instance of `kind: Foo` is allowed per namespace). I suspect that it would not be significantly more work to cleanly handle multiple instances of each easy stack custom resource being created in one namespace, and also suspect that the user experience would be fragile/surprising/weird if we _did not_ support this.

Reply:
From comments and our chat yesterday, the direction we're taking is expose a field to accept a secret ref and support multiple CR instances.